File: backend/app/models/history.py
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, JSON, Float, Enum
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from ..database import Base
import enum

class ActionType(str, enum.Enum):
    CALCULATE = "calculate"
    SAVE = "save"
    DELETE = "delete"
    VIEW = "view"

class History(Base):
    __tablename__ = "history"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    scenario_id = Column(Integer, ForeignKey("scenarios.id"), nullable=True)
    action = Column(Enum(ActionType), nullable=False)
    details = Column(JSON, default=dict)
    calculation_time_ms = Column(Float, nullable=True)  
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    user = relationship("User", back_populates="history")
    scenario = relationship("Scenario", back_populates="history")

    def __repr__(self):
        return f"History(id={self.id}, user_id={self.user_id}, action={self.action})"

File: frontend/src/types/history.ts
export type ActionType = 'calculate' | 'save' | 'delete' | 'view'

export interface HistoryRecord {
  id: number
  user_id: number
  scenario_id: number | null
  scenario_name: string | null
  action: ActionType
  details: Record<string, unknown>
  calculation_time_ms: number | null
  created_at: string
}

export interface HistoryList {
  items: HistoryRecord[]
  total: number
  page: number
  size: number
  pages: number
}
